src/main/java/org/springframework/retry/annotation/RecoverAnnotationRecoveryHandler.java

```diff
@@ -76,8 +79,19 @@ public T recover(Object[] args, Throwable cause) {
 		boolean methodAccessible = method.isAccessible();
 		try {
 			ReflectionUtils.makeAccessible(method);
+			RetryContext context = RetrySynchronizationManager.getContext();
+			Object proxy = null;
+			if (context != null) {
+				proxy = context.getAttribute("___proxy___");
+				if (AopUtils.isJdkDynamicProxy(proxy)) {
+					proxy = null;
+				}
+			}
+			if (proxy == null) {
+				proxy = this.target;
```

Comment:
What is the problem if we always propagate a proxy to this instance not just target?

Reply:
1. There are some tests that call this directly, so no proxy is added to the context.
2. If the proxy is a JDK proxy we can't invoke the recoverer because the method is the instance method, not the interface method.

```
java.lang.IllegalArgumentException: object is not an instance of declaring class
```

Reply:
1. I am not sure if designing a change not to break existing tests is such a good idea. Maybe you should updatte the tests instead.
2. Even if the proxy is a JDK proxy, a proxied interface method could point to a method in the target instance which is annotated by `@Recover`. Can you explain what exactly does not work there? Is it just difficult to discover the target method or is there some other basic problem which I simply don't understand, being an AspectJ expert but a relative SpringAOP layman?

Reply:
1. Good point; I was just being lazy to get a quick fix out.
2. Yes, we can add code to find the corresponding method on the proxy (if it exists). The problem is it won't solve every use case because the proxied interfaces may not expose the recover method. So we'd still need to handle the case where we call the recover method directly on the target instead of the proxy.

Reply:
2. Yes, I know. In this case...
```java
public interface Foo {
  String bar();
```
```java
@Component public class FooImpl {
  @Override public String bar() { return "bar"; }
  @Recover public int notFromInterface() { return 42; }
}
```
... of course it will not work, because `notFromInterface` is not exposed by a JDK proxy.

But this would be detectable:

```java
@Component public class FooImpl {
  @Override @Recover public String bar() { return "bar"; }
}
```